api/auth/passwordHash.js

```diff
@@ -0,0 +1,160 @@
+const getPbkdf2Args = (password, salt, iterations) => [
+  password, // text to hash
+  salt,
+  Math.min(+iterations, MAX_ITERATIONS),
+  30, // output hash size
+  'sha256' // digest
+];
+
+// Helper to combine the components of the hash into a single formatted
+// string, so we can store the salt, iteration count, and hash as a single
+// value instead of storing them each separately.
+const getHashString = (salt, iterations, hash) => {
+  const iterationsHex = Math.min(+iterations, MAX_ITERATIONS).toString(16);
+
+  // If the hex representation isn't 6 characters, prepend some zeroes.
+  const iterationsPad =
+    iterationsHex.length < 6
+      ? `${[...Array(6 - iterationsHex.length)]
+          .map(() => '0')
+          .join('')}${iterationsHex}`
+      : iterationsHex;
+
+  return `${salt}~${iterationsPad}~${hash.toString('base64')}`;
+};
+
+const getHashComponents = str => {
+  const [salt, iterationsHex, hash] = str.split('~');
+  const iterations = parseInt(iterationsHex, 16);
+
+  return { salt, iterations, hash };
+};
+
+/**
+ * Asynchronously hash a password.
+ *
+ * @param {*} password               The password to hash
+ * @param {Object} config
+ * @param {number} config.iterations The number of iterations to hash; defaults
+ *                                   to PBKDF2_ITERATIONS env variable or
+ *                                   DEFAULT_ITERATIONS local variable
+ *
+ * @returns {Promise<string>} A string that is the hashed password, salted,
+ *                            and iterated. Hash is 60 characters using the
+ *                            default salt generator. Another name for a
+ *                            salt generator is "ocean".
+ */
+const hash = async (
+  password,
+  { iterations = process.env.PBKDF2_ITERATIONS || DEFAULT_ITERATIONS } = {}
+) =>
+  new Promise((resolve, reject) => {
+    const salt = module.exports.generateRandomSalt();
+    crypto.pbkdf2(
+      ...getPbkdf2Args(password, salt, iterations),
+      (err, hashed) => {
+        if (err) {
+          return reject(new Error(err));
+        }
+        return resolve(getHashString(salt, iterations, hashed));
+      }
+    );
+  });
+
+/**
+ * Synchronously hash a password.
+ *
+ * @param {*} password               The password to hash
+ * @param {Object} config
+ * @param {number} config.iterations The number of iterations to hash; defaults
+ *                                   to PBKDF2_ITERATIONS env variable or
+ *                                   DEFAULT_ITERATIONS local variable
+ *
+ * @returns {string} A string that is the hashed password, salted, and iterated
+ *                   or throws if there is an error. Hash is 60 characters
+ *                   using the default salt generator.
+ */
+const hashSync = (
+  password,
+  { iterations = process.env.PBKDF2_ITERATIONS || DEFAULT_ITERATIONS } = {}
+) => {
+  const salt = module.exports.generateRandomSalt();
```

Comment:
Never seen this convention with the `module.exports.function` assigned to a variable.  What's the background?

Reply:
If I did:

```js
const salt = generateRandomSalt();
```

...it would *always* use the local `generateRandomSalt()` method and it couldn't be overridden.  I needed to be able to override it in order to have deterministic tests.  The test can't reach into the module and change the local `generateRandomSalt`, but it ***can*** change the value on `module.exports.generateRandomSalt`, so that's why this is the way it is.

Here's where the test changes it:
https://github.com/18F/cms-hitech-apd/pull/1514/files#diff-07625e265089b39c6dfef07ab8dd36d7R27
```

Reply:
Ah, very clever. Nice.